examples/reader-writer.cc

```diff
@@ -0,0 +1,426 @@
+    // Create a primitive node named 'double_field' with type:DOUBLE, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("double_field", parquet::Repetition::REQUIRED, parquet::Type::DOUBLE, parquet::LogicalType::NONE));
+    // Create a primitive node named 'ba_field' with type:BYTE_ARRAY, repetition:OPTIONAL
+    fields.push_back(parquet::schema::PrimitiveNode::Make("ba_field", parquet::Repetition::OPTIONAL, parquet::Type::BYTE_ARRAY, parquet::LogicalType::NONE));
+    // Create a primitive node named 'flba_field' with type:FIXED_LEN_BYTE_ARRAY, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("flba_field", parquet::Repetition::REQUIRED, parquet::Type::FIXED_LEN_BYTE_ARRAY, parquet::LogicalType::NONE, FIXED_LENGTH));
+
+    // Create a GroupNode named 'schema' using the primitive nodes defined above
+    // This GroupNode is the root node of the schema tree
+    schema_ = std::static_pointer_cast<parquet::schema::GroupNode>(parquet::schema::GroupNode::Make("schema", parquet::Repetition::REQUIRED, fields));
+  }
+
+  parquet::ParquetFileWriter* GetWriter() {
+    if (!writer_) {
+      // Create a parquet file writer using the schema and the output stream.
+      writer_ = parquet::ParquetFileWriter::Open(ostream_, schema_);
+    }
+    return writer_.get();
+  }
+
+  void WriteToFile(std::string filename) {
+    std::ofstream parquetFile;
+    parquetFile.open(filename, std::ios::out | std::ios::binary);
+    std::shared_ptr<parquet::Buffer> buffer = ostream_->GetBuffer();
+    parquetFile.write(reinterpret_cast<const char*>(buffer->data()), buffer->size());
+    parquetFile.close();
+  }
+
+ private:
+  std::shared_ptr<parquet::schema::GroupNode> schema_;
+  std::shared_ptr<parquet::InMemoryOutputStream> ostream_;
+  std::unique_ptr<parquet::ParquetFileWriter> writer_;
+};
+
+int main(int argc, char** argv) {
+
+/**********************************************************************************
+                           PARQUET WRITER EXAMPLE
+**********************************************************************************/
+  // parquet::REQUIRED fields do not need definition and repetition level values
+  // parquet::OPTIONAL fields require only definition level values
+  // parquet::REPEATED fields require both definition and repetition level values
+  try {
+    // Create a ParquetWriter instance
+    auto parquet_writer = std::unique_ptr<ParquetWriter>(new ParquetWriter());
+
+    // Setup the schema
+    parquet_writer->SetupSchema();
+
+    // Get a handle to the ParquetFileWriter
+    parquet::ParquetFileWriter* file_writer = parquet_writer->GetWriter();
+
+    // Append a RowGroup with a specific number of rows.
+    parquet::RowGroupWriter* rgWriter = file_writer->AppendRowGroup(NUM_ROWS_PER_ROW_GROUP);
+
+    // Write the Bool column 
+    parquet::ColumnWriter* colWriter = rgWriter->NextColumn();
```

Comment:
Yes. I would like the user to cast to them to actually have typed `Write` calls. In the end you either have to cast your data pointer with `reinterpret_cast<uint8_t*>` or the `ColumnWriter` to a `TypedColumnWriter`.

Using the `WriteAllValues` will also prohibit some future features like "I want to write an uint8_t to an int32_t physical column`". We can support this via differently typed `WriteBatch` call, e.g. we do this already for `const bool*` and `const std::vector<bool>&` for Boolean columns.